include/pacmod2/pacmod2_node.hpp

```diff
@@ -0,0 +1,149 @@
+
+  /// \brief Callback from transition to "error" state.
+  /// \param[in] state The current state that the node is in.
+  LNI::CallbackReturn on_error(const lc::State & state) override;
+
+private:
+  void initializeBrakeMotorRptApi();
+  void initializeSteeringMotorRptApi();
+  //void initializeWiperApi();
+  void initializeHeadlightApi();
+  void initializeHornApi();
+  void initializeWheelSpeedApi();
+  //void initializeParkingBrakeRptApi();
+  void initializeHazardLightApi();
+
+  void initializeApiForMsg(uint32_t msg_can_id);
+
+  void callback_can_tx(const can_msgs::msg::Frame::SharedPtr msg);
+  void callback_global_cmd(const pacmod2_msgs::msg::GlobalCmd::SharedPtr msg);
+  void callback_accel_cmd(const pacmod2_msgs::msg::SystemCmdFloat::SharedPtr msg);
+  void callback_brake_cmd(const pacmod2_msgs::msg::SystemCmdFloat::SharedPtr msg);
+  void callback_hazard_lights_cmd(const pacmod2_msgs::msg::SystemCmdBool::SharedPtr msg);
+  void callback_headlight_cmd(const pacmod2_msgs::msg::SystemCmdInt::SharedPtr msg);
+  void callback_horn_cmd(const pacmod2_msgs::msg::SystemCmdBool::SharedPtr msg);
+  void callback_shift_cmd(const pacmod2_msgs::msg::SystemCmdInt::SharedPtr msg);
+  void callback_steering_cmd(const pacmod2_msgs::msg::PositionWithSpeed::SharedPtr msg);
+  void callback_turn_cmd(const pacmod2_msgs::msg::SystemCmdInt::SharedPtr msg);
+  void callback_wiper_cmd(const pacmod2_msgs::msg::SystemCmdInt::SharedPtr msg);
+
+  template<class T>
+  void lookup_and_encode(const unsigned int & can_id, const T & msg)
+  {
+    auto cmd = can_subs_.find(can_id);
+
+    if (cmd != can_subs_.end()) {
+      cmd->second.second->setData(Pacmod2RxRosMsgHandler::unpackAndEncode(can_id, msg));
+    } else {
+      RCLCPP_WARN(
+        this->get_logger(),
+        "Received a command message for ID 0x%x for which we do not have an encoder.",
+        can_id);
+    }
+  }
+
+  void publish_cmds();
+  //void set_enable(bool enable);
+
+  static constexpr auto SEND_CMD_INTERVAL = std::chrono::milliseconds(33);
+  static constexpr auto INTER_MSG_PAUSE = std::chrono::milliseconds(1);
+
+  std::string frame_id_;
+  unsigned int dbc_major_version_;
+  Pacmod2TxRosMsgHandler tx_handler_;
+  std::map<unsigned int, std::tuple<bool, bool, bool>> system_statuses;
+
+  std::shared_ptr<rclcpp::TimerBase> system_statuses_timer_;
+  std::shared_ptr<lc::LifecyclePublisher<can_msgs::msg::Frame>> pub_can_rx_;
+  std::unordered_map<unsigned int, std::shared_ptr<lc::LifecyclePublisherInterface>> can_pubs_;
+  std::shared_ptr<lc::LifecyclePublisher<std_msgs::msg::Bool>> pub_enabled_;
+  std::shared_ptr<lc::LifecyclePublisher<
+      pacmod2_msgs::msg::AllSystemStatuses>> pub_all_system_statuses_;
+
+  std::shared_ptr<rclcpp::Subscription<can_msgs::msg::Frame>> sub_can_tx_;
+  std::unordered_map<unsigned int,
+    std::pair<std::shared_ptr<rclcpp::SubscriptionBase>,
+    std::shared_ptr<LockedData>>> can_subs_;
+
+  std::shared_ptr<std::thread> pub_thread_;
```

Comment:
if this pub_thread_ is not shared with anyone else, use std::unique_ptr instead.

Reply:
done